source/convergence_test.cc

```diff
@@ -0,0 +1,684 @@
+                    ExcLowerRange(n_spatial_cycles, 0));
+    }
+    else
+      AssertThrow(false,
+                  ExcMessage("Unexpected identifier for the type of"
+                             " of convergence test."));
+  }
+  prm.leave_subsection();
+}
+
+
+
+template<typename Stream>
+Stream& operator<<(Stream &stream, const ConvergenceTestParameters &prm)
+{
+  internal::add_header(stream);
+  internal::add_line(stream, "Convergence test parameters");
+  internal::add_header(stream);
+
+  switch (prm.test_type)
+  {
+    case ConvergenceTestType::spatial:
+      internal::add_line(stream, "Convergence test type", "spatial");
+      break;
+    case ConvergenceTestType::temporal:
+      internal::add_line(stream, "Convergence test type", "temporal");
+      break;
+    default:
+      AssertThrow(false, ExcMessage("Unexpected identifier for the type of"
+                               " of convergence test."));
+      break;
+  }
+
+  internal::add_line(stream,
+                     "Number of spatial convergence cycles",
+                     prm.n_spatial_cycles);
+
+  internal::add_line(stream,
+                     "Number of temporal convergence cycles",
+                     prm.n_temporal_cycles);
+
+  internal::add_line(stream,
+                     "Time-step reduction factor",
+                     prm.step_size_reduction_factor);
+
+  internal::add_header(stream);
+
+  return (stream);
+}
+
+
+ConvergenceTestData::ConvergenceTestData(const ConvergenceTestType &type)
+:
+type(type),
+level(0)
+{}
+
+template<int dim, int spacedim>
+void ConvergenceTestData::update_table
+(const DoFHandler<dim, spacedim> &dof_handler,
+ const double step_size,
+ const std::map<typename VectorTools::NormType, double> &error_map)
+{
+  table.add_value("step size", step_size);
+  step_size_specified = true;
+
+  update_table(dof_handler, error_map);
+}
+
+
+template<int dim, int spacedim>
+void ConvergenceTestData::update_table
+(const DoFHandler<dim, spacedim> &dof_handler,
+ const std::map<typename VectorTools::NormType, double> &error_map)
+{
+  table.add_value("level", level);
+  level += 1;
+
+  const types::global_dof_index n_dofs{dof_handler.n_dofs()};
+  table.add_value("n_dofs", n_dofs);
+
+  const Triangulation<dim, spacedim> &tria{dof_handler.get_triangulation()};
+  const types::global_cell_index n_cells{tria.n_global_active_cells()};
+  table.add_value("n_cells", n_cells);
+
+  const double h_max{GridTools::maximal_cell_diameter(tria)};
+  table.add_value("h_max", h_max);
+  h_max_specified = true;
+
+  std::string column;
+  unsigned int dimension{spacedim};
+  bool evaluate_convergence{false};
+
+  switch (type)
+  {
+    case ConvergenceTestType::spatial:
+      column.assign("h_max");
+      evaluate_convergence = true;
+      break;
+    case ConvergenceTestType::temporal:
+      column.assign("step size");
+      dimension = 1;
+      evaluate_convergence = true;
+      break;
+    case ConvergenceTestType::spatio_temporal:
```

Comment:
I see some `switch` scopes considering this third options and other who does not take it in consideration. Is this by design?

Reply:
For a spatio-temporal convergence test in `ConvergenceTestData` the only difference is that the convergence rates are not evaluated.